agent/acs/handler/acs_handler.go

```diff
@@ -189,70 +189,59 @@ func NewSession(
-			if shouldReconnectWithoutBackoff(acsError) {
-				// If ACS closed the connection, there's no need to backoff,
-				// reconnect immediately
-				seelog.Infof("ACS Websocket connection closed for a valid reason: %v", acsError)
-				acsSession.backoff.Reset()
-				sendEmptyMessageOnChannel(connectToACS)
-			} else {
-				// Disconnected unexpectedly from ACS, compute backoff duration to
-				// reconnect
-				reconnectDelay := acsSession.computeReconnectDelay(isInactiveInstance)
-				seelog.Infof("Reconnecting to ACS in: %s", reconnectDelay.String())
-				waitComplete := acsSession.waitForDuration(reconnectDelay)
-				if waitComplete {
-					// If the context was not cancelled and we've waited for the
-					// wait duration without any errors, send the message to the channel
-					// to reconnect to ACS
-					seelog.Info("Done waiting; reconnecting to ACS")
-					sendEmptyMessageOnChannel(connectToACS)
-				} else {
-					// Wait was interrupted. We expect the session to close as canceling
-					// the session context is the only way to end up here. Print a message
-					// to indicate the same
-					seelog.Info("Interrupted waiting for reconnect delay to elapse; Expect session to close")
-				}
-			}
 		case <-acsSession.ctx.Done():
 			// agent is shutting down, exiting cleanly
 			return nil
+		default:
 		}
 
+		// Session with ACS was stopped with some error, start processing the error
+		isInactiveInstance := isInactiveInstanceError(acsError)
+		if isInactiveInstance {
+			// If the instance was deregistered, send an event to the event stream
+			// for the same
+			seelog.Debug("Container instance is deregistered, notifying listeners")
+			err := acsSession.deregisterInstanceEventStream.WriteToEventStream(struct{}{})
+			if err != nil {
+				seelog.Debugf("Failed to write to deregister container instance event stream, err: %v", err)
+			}
+		}
+		if shouldReconnectWithoutBackoff(acsError) {
+			// If ACS closed the connection, there's no need to backoff,
+			// reconnect immediately
+			seelog.Infof("ACS Websocket connection closed for a valid reason: %v", acsError)
+			acsSession.backoff.Reset()
+		} else {
+			// Disconnected unexpectedly from ACS, compute backoff duration to
+			// reconnect
+			reconnectDelay := acsSession.computeReconnectDelay(isInactiveInstance)
+			seelog.Infof("Reconnecting to ACS in: %s", reconnectDelay.String())
+			waitComplete := acsSession.waitForDuration(reconnectDelay)
+			if waitComplete {
+				// If the context was not cancelled and we've waited for the
+				// wait duration without any errors, send the message to the channel
+				// to reconnect to ACS
+				seelog.Info("Done waiting; reconnecting to ACS")
+			} else {
+				// Wait was interrupted. We expect the session to close as canceling
+				// the session context is the only way to end up here. Print a message
+				// to indicate the same
+				seelog.Info("Interrupted waiting for reconnect delay to elapse; Expect session to close")
+				return nil
```

Comment:
Note that this isn't equivalent to the prev revision. This exits without waiting for the context to get canceled. The equivalent version would be do `<- ctx.Done()` before returning

Reply:
It actually is. In the sense that Done() being signaled is the only way that false is returned. We could check it again here if you want but it isn't any different.

Reply:
you are right